src/config.rs

```diff
@@ -777,40 +779,47 @@ impl<'a> Cfg<'a> {
 
     // Returns a Toolchain matching the given ToolchainDesc, installing it and
     // the given components and targets if they aren't already installed.
-    async fn ensure_installed(
+    #[tracing::instrument(level = "trace", err(level = "trace"), skip_all)]
+    pub(crate) async fn ensure_installed(
         &self,
         toolchain: ToolchainDesc,
         components: Vec<String>,
         targets: Vec<String>,
         profile: Option<Profile>,
-    ) -> Result<Toolchain<'_>> {
+        verbose: bool,
+    ) -> Result<(UpdateStatus, Toolchain<'_>)> {
+        if verbose {
+            (self.notify_handler)(Notification::LookingForToolchain(&toolchain));
+        }
         let components: Vec<_> = components.iter().map(AsRef::as_ref).collect();
         let targets: Vec<_> = targets.iter().map(AsRef::as_ref).collect();
-        let toolchain = match DistributableToolchain::new(self, toolchain.clone()) {
+        let profile = profile.map_or_else(|| self.get_profile(), Ok)?;
```

Comment:
This commit has a lot going on. I'd probably recommend splitting it up in set of smaller commits, maybe like this:

* Add `tracing::instrument` attribute here
* Add notifications, contingent on `verbose` flag
* More sophisticated profile selection if not specified
* Yield status from `ensure_installed()`

For the profile thing, IMO `map_or_else()` isn't great for this case, would be clearer to use an explicit `match` here:

```rust
let profile = match profile {
    Some(profile) => profile,
    None => self.get_profile()?,
};
```